lib/encodings/base64.fz

```diff
@@ -0,0 +1,155 @@
+# Base64 encoding and decoding as defined in RFC 4648
+public base64 is
+
+  # allows redefinition for base64url
+  module get_alphabet => "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".as_codepoints.as_array
+  module get_char_62 u8 => 43 # +
+  module get_char_63 u8 => 47 # /
+  module get_encoding_name => "base64"
+
+  # NYI: compiler does currently not optimize features without arguments to treat them like fields
+  char_62_ascii := get_char_62
+  char_63_ascii := get_char_63
+  alphabet := get_alphabet
+  encoding_name := get_encoding_name
+
+  # Encodes a given byte sequence in base64, output length is padded to multiple of 4
+  public encode (data Sequence u8) String =>
+
+    # 0 -> 0; 1 -> 2; 2 -> 1
+    pad_size := (3 - data.count%3)%3     # length of the padding: number of padded zero bytes in input as well as number of '=' in output
+    in_len := data.count + pad_size      # length of zero padded input data
+
+    # this guarantees that input length is always multiple of 3
+    input := array u8 in_len i->if i < data.count then data[i] else 0
+
+    for b0 := 0, b0 + 3
+        b1 := 1, b1 + 3
+        b2 := 2, b2 + 3
+        res := "", res + block # the encoded input data
+    while b2 < in_len
+      # get 3 bytes (24 bits) and convert them into a number with 24 bits
+      n u32 := (input[b0].as_u32 << 16) + (input[b1].as_u32 << 8) + input[b2].as_u32
```

Comment:
This must be optimized! `input[b0]` requires time in `O(b0)`, so the loop is in `O(in_len²)`!

My suggestion: Do a loop as follows
```
for 
  res := "", res + block # the encoded input data
  i in 0..
  b in data
  last_n := 0, i %% 3 ?  0 : n
  n := last_n << 8 + b 
do 
  block := if i %% 3 then
              n0 => ..
              ...
              block := alphabet[n0]...
            else
              ""
else
   ...
    
```

Reply:
I tried this, it gave me a bit of a headache, but it works.

The code is not shorter though, but maybe I made it too complicated.
I don't know if there is a difference in performance (in the old version the input is converted into an array so that should not matter). However, I think it's less understandable, at least for new users.

See c04016f132947328c73c4250da020bf62a05380e

What are your thoughts?